src/System.Text.Json/tests/Serialization/ExtensionDataTests.cs

```diff
@@ -353,21 +354,129 @@ private class ClassWithReference
+
+            var parent = new ParentClassWithObject { Text = "Hello World" };
+            parent.ExtensionData.Add("SpecialInformation", "I am parent class");
+            parent.Child = child;
+
+            // The extension data is based on the raw strings added above and not JsonElement.
+            Assert.Equal("Hello World", parent.Text);
+            Assert.IsType<string>(parent.ExtensionData["SpecialInformation"]);
+            Assert.Equal("I am parent class", (string)parent.ExtensionData["SpecialInformation"]);
+            Assert.Equal(2, parent.Child.Number);
+            Assert.IsType<string>(parent.Child.ExtensionData["SpecialInformation"]);
+            Assert.Equal("I am child class", (string)parent.Child.ExtensionData["SpecialInformation"]);
+
+            // Round-trip and verify. Extension data is now based on JsonElement.
+            string json = JsonSerializer.Serialize(parent);
+            parent = JsonSerializer.Deserialize<ParentClassWithObject>(json);
+
+            Assert.Equal("Hello World", parent.Text);
+            Assert.IsType<JsonElement>(parent.ExtensionData["SpecialInformation"]);
+            Assert.Equal("I am parent class", ((JsonElement)parent.ExtensionData["SpecialInformation"]).GetString());
+            Assert.Equal(2, parent.Child.Number);
+            Assert.IsType<JsonElement>(parent.Child.ExtensionData["SpecialInformation"]);
+            Assert.Equal("I am child class", ((JsonElement)parent.Child.ExtensionData["SpecialInformation"]).GetString());
+        }
 
-            ClassWithReference obj = JsonSerializer.Deserialize<ClassWithReference>(json);
-            Assert.IsType<JsonElement>(obj.MyOverflow["MyIntMissing"]);
-            Assert.Equal(1, obj.MyOverflow.Count);
-            Assert.Equal(2, obj.MyOverflow["MyIntMissing"].GetInt32());
+        private class ParentClassWithJsonElement
+        {
+            public string Text { get; set; }
+
+            public List<ChildClassWithJsonElement> Children { get; set; } = new List<ChildClassWithJsonElement>();
 
-            ClassWithExtensionProperty child = obj.MyReference;
-            Assert.IsType<JsonElement>(child.MyOverflow["MyIntMissingChild"]);
-            Assert.IsType<JsonElement>(child.MyOverflow["MyIntMissingChild"]);
-            Assert.Equal(1, child.MyOverflow.Count);
-            Assert.Equal(3, child.MyOverflow["MyIntMissingChild"].GetInt32());
+            [JsonExtensionData]
+            public Dictionary<string, JsonElement> ExtensionData { get; set; } = new Dictionary<string, JsonElement>();
+        }
+
+        private class ChildClassWithJsonElement
+        {
+            public int Number { get; set; }
+
+            [JsonExtensionData]
+            public Dictionary<string, JsonElement> ExtensionData { get; set; } = new Dictionary<string, JsonElement>();
+        }
+
+        [Fact]
+        public static void NestedClassWithJsonElementExtensionDataProperty()
```

Comment:
Just to verify, will all these new tests fail without the source change?

Reply:
Yes